File: Structural/Flyweight/JavaScript/characterFactory.js
// characterFactory.js

const FlyweightCharacter = require("./flyweightCharacter");

class CharacterFactory {
  static characterPool = new Map();

  static getCharacter(symbol) {
    if (!this.characterPool.has(symbol)) {
      this.characterPool.set(symbol, new FlyweightCharacter(symbol));
    }
    return this.characterPool.get(symbol);
  }

  static getPoolSize() {
    return this.characterPool.size;
  }
}

module.exports = CharacterFactory;

File: Structural/Flyweight/Java/CharacterFactory.java
// CharacterFactory.java

import java.util.HashMap;
import java.util.Map;

public class CharacterFactory {
  private static final Map<Character, FlyweightCharacter> characterPool = new HashMap<>();

  public static FlyweightCharacter getCharacter(char symbol) {
    characterPool.putIfAbsent(symbol, new FlyweightCharacter(symbol));
    return characterPool.get(symbol);
  }

  public static int getPoolSize() {
    return characterPool.size();
  }
}
